p2p/host.go

```diff
@@ -51,154 +72,225 @@ func newHost(ctx context.Context, cfg *Config) (*host, error) {
 		return nil, err
 	}
 
+	// use default protocol if none provided
 	protocolId := protocol.ID(cfg.ProtocolId)
 	if protocolId == "" {
 		protocolId = DefaultProtocolId
 	}
 
+	// create datastore and dht service
 	dstore := dsync.MutexWrap(ds.NewMapDatastore())
 	dht := kaddht.NewDHT(ctx, h, dstore)
 
-	// wrap the host with routed host so we can look up peers in DHT
+	// wrap the host with routed host so we can look up peers in dht
 	h = rhost.Wrap(h, dht)
 
-	// build host multiaddress
-	hostAddr, err := ma.NewMultiaddr(fmt.Sprintf("/p2p/%s", h.ID().Pretty()))
+	// use "p2p" for multiaddress format
+	ma.SwapToP2pMultiaddrs()
+
+	// create host multiaddress including host "p2p" id
+	address, err := ma.NewMultiaddr(fmt.Sprintf("/p2p/%s", h.ID()))
 	if err != nil {
 		return nil, err
 	}
 
+	// format bootstrap nodes
 	bootstrapNodes, err := stringsToPeerInfos(cfg.BootstrapNodes)
 	if err != nil {
 		return nil, err
 	}
 
+	// TODO: store status in peer metadata
+	peerStatus := make(map[peer.ID]bool)
+
 	return &host{
 		ctx:         ctx,
 		h:           h,
-		hostAddr:    hostAddr,
 		dht:         dht,
 		bootnodes:   bootstrapNodes,
-		protocolId:  protocolId,
 		noBootstrap: cfg.NoBootstrap,
 		noMdns:      cfg.NoMdns,
+		address:     address,
+		protocolId:  protocolId,
+		peerStatus:  peerStatus, // TODO: store status in peer metadata
 	}, nil
 
 }
 
-// bootstrap connects to all provided bootnodes unless noBootstrap is true
+// bootstrap connects the host to the configured bootnodes
 func (h *host) bootstrap() {
 	if len(h.bootnodes) == 0 && !h.noBootstrap {
-		log.Debug("no peers to bootstrap to")
+		log.Error(
+			"bootstrap",
+			"error", "no bootnodes defined and bootstrap enabled",
+		)
 	}
-
+	// loop through bootnode peers
 	for _, peerInfo := range h.bootnodes {
-		log.Debug("Attempting to dial peer", "id", peerInfo.ID, "addrs", peerInfo.Addrs)
-		err := h.h.Connect(context.Background(), peerInfo)
+		log.Trace(
+			"bootstrap",
+			"host", h.id(),
+			"peer", peerInfo.ID,
+		)
+		// connect to each peer
+		err := h.connect(peerInfo)
 		if err != nil {
-			log.Debug("failed to dial bootstrap peer", "err", err)
+			log.Error("connect", "error", err)
```

Comment:
Hmm... I can revert but `err` will provide the error message for why `connect` failed. Also, will `connect` throw any other errors rather than just a failed dial error?

Reply:
Fixed.